File: v1/src/main/java/com/api/v1/services/AllPaymentsRetrievalServiceImpl.java
package com.api.v1.services;

import com.api.v1.domain.PaymentRepository;
import com.api.v1.dtos.PaymentResponseDto;
import com.api.v1.exceptions.EmptyPaymentEntityException;
import com.api.v1.mappers.PaymentResponseMapper;
import com.api.v1.utils.PaymentFinderUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
class AllPaymentsRetrievalServiceImpl implements AllPaymentsRetrievalService {

    @Autowired
    private PaymentFinderUtil paymentFinderUtil;

    @Autowired
    private PaymentRepository repository;

    @Override
    public Flux<PaymentResponseDto> findAll() {
        return repository
                .findAll()
                .hasElements()
                .flatMapMany(hasElements -> {
                    if (!hasElements) return Mono.error(EmptyPaymentEntityException::new);
                    return repository
                            .findAll()
                            .flatMap(payment -> Mono.just(
                                    new PaymentResponseMapper(
                                            payment,
                                            payment.getCustomer(),
                                            payment.getCar(),
                                            payment.getCreditCard())
                                            .map()
                            ));
                });
    }

    @Override
    public Mono<PaymentResponseDto> findByOrderNumber(String orderNumber) {
        return paymentFinderUtil
                .find(orderNumber)
                .flatMap(payment -> Mono.just(
                        new PaymentResponseMapper(
                                payment,
                                payment.getCustomer(),
                                payment.getCar(),
                                payment.getCreditCard())
                                .map()
                ));
    }
}

File: v2/src/main/kotlin/com/api/v2/payment/services/AllPaymentsRetrievalServiceImpl.kt
package com.api.v2.payment.services

import com.api.v2.payment.domain.PaymentRepository
import com.api.v2.payment.dtos.PaymentResponseDto
import com.api.v2.payment.exceptions.PaymentEntityEmptyException
import com.api.v2.payment.mappers.PaymentResponseMapper
import com.api.v2.payment.utils.PaymentFinderUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class AllPaymentsRetrievalServiceImpl: AllPaymentsRetrievalService {

    @Autowired
    lateinit var paymentFinderUtil: PaymentFinderUtil

    @Autowired
    lateinit var paymentRepository: PaymentRepository

    override suspend fun findAll(): Flow<PaymentResponseDto> {
        return withContext(Dispatchers.IO) {
            if (paymentRepository.findAll().count() == 0) {
                throw PaymentEntityEmptyException()
            }
            paymentRepository
                .findAll()
                .map { e -> PaymentResponseMapper.map(e, e.customer, e.car, e.creditCard) }
        }

    }

    override suspend fun findByOrderNumber(orderNumber: String): PaymentResponseDto {
        return withContext(Dispatchers.IO) {
            val payment = paymentFinderUtil.find(orderNumber)
            PaymentResponseMapper.map(
                payment,
                payment.customer,
                payment.car,
                payment.creditCard)
        }
    }

}